src/lightning/pytorch/utilities/restricted_classmethod.py

```diff
@@ -0,0 +1,31 @@
+from typing import Callable, Concatenate, Generic, ParamSpec, TypeVar
+
+
+class RestrictedClassmethodError(Exception):
+    """
+    This exception raised when a `restricted_classmethod` is invoked on an instance
+    instead of a class type.
+    """
+
+
+_T = TypeVar("_T")
+_P = ParamSpec("_P")
+_R_co = TypeVar("_R_co", covariant=True)
```

Comment:
This level of typing is more complex than what the average code reader would understand. Mind adding some comments that explain why we need all these?